File: Data-Structure/Nodejs/LeetCode by Category/Graphs/Topological Sort 拓扑排序/210. Course Schedule II.js
//210. Course Schedule II
/*
There are a total of numCourses courses you have to take, labeled from 0 to numCourses - 1. You are given an array prerequisites where prerequisites[i] = [ai, bi] indicates that you must take course bi first if you want to take course ai.

For example, the pair [0, 1], indicates that to take course 0 you have to first take course 1.
Return the ordering of courses you should take to finish all courses. If there are many valid answers, return any of them. If it is impossible to finish all courses, return an empty array.

Example 1:
Input: numCourses = 2, prerequisites = [[1,0]]
Output: [0,1]
Explanation: There are a total of 2 courses to take. To take course 1 you should have finished course 0. So the correct course order is [0,1].

Example 2:
Input: numCourses = 4, prerequisites = [[1,0],[2,0],[3,1],[3,2]]
Output: [0,2,1,3]
Explanation: There are a total of 4 courses to take. To take course 3 you should have finished both courses 1 and 2. Both courses 1 and 2 should be taken after you finished course 0.
So one correct course order is [0,1,2,3]. Another correct ordering is [0,2,1,3].
*/
//https://zhuanlan.zhihu.com/p/135094687
{
    /**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
    var findOrder = function (numCourses, prerequisites) {
        let res = Array(numCourses).fill(0);
        let indegree = Array(numCourses).fill(0); //入度

        // get the indegree for each course
        for (let pre of prerequisites) {
            indegree[pre[0]]++;
        }
        // put courses with indegree == 0 to queue
        let queue = [];
        for (let i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                queue.push(i);
            }
        }
        //console.log(queue);
        // execute the course
        let i = 0;
        while (queue.length) {
            let curr = queue.shift();
            res[i++] = curr;
            // remove the pre = curr
            for (let pre of prerequisites) {
                if (pre[1] == curr) {
                    indegree[pre[0]]--;
                    if (indegree[pre[0]] == 0) {
                        queue.push(pre[0]);
                    }
                }
            }
        }

        return i === numCourses ? res : [];
    };
    //Course Schedule Medium
    //Alien Dictionary Hard
    //Minimum Height Trees Medium
    //Sequence Reconstruction Medium
    //Course Schedule III Hard
    //Parallel Courses Medium

}
//https://leetcode.com/problems/course-schedule-ii/discuss/942920/DFS-Topological-sort-JS-Solution
{
    // DFS
    var findOrder = function (numCourses, prerequisites) {
        let adjList = new Array(numCourses).fill(0).map(() => []);
        let visited = new Array(numCourses).fill(false);
        let inDegree = new Array(numCourses).fill(0);
        for (let [course, preCourse] of prerequisites) {
            adjList[preCourse].push(course);
            inDegree[course]++;
        }
        // We can take courses that have no prerequisite or whose pre-requisites have already been taken
        let res = [];
        for (let i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0 && !visited[i]) {
                dfs(i);
            }
        }

        return res.length === numCourses ? res : [];

        function dfs(node) {
            res.push(node);
            visited[node] = true;
            for (let next of adjList[node]) {
                inDegree[next]--;
                if (inDegree[next] == 0 && !visited[next]) {
                    dfs(next);
                }
            }
        }
        // Time Complexity: O(V + E), if there is a valid answer, we visit every node and all of its neighbors (those that current node is pointing to)
        // Space Complexity: O(V + E), for adjacency list  
    };
}
{
    var findOrder = function (numCourses, prerequisites) {
        const adjacencyList = [];
        const inDegree = new Array(numCourses).fill(0);
        for (i = 0; i < numCourses; i++) {
            adjacencyList.push([]);
        }
        for (i = 0; i < prerequisites.length; i++) {
            const req = prerequisites[i];
            adjacencyList[req[1]].push(req[0]);
            inDegree[req[0]]++;
        }

        const q = [];
        const order = [];

        for (i = 0; i < numCourses; i++) {
            if (inDegree[i] === 0) {
                q.push(i);
            }
        }

        while (q.length) {
            const node = q.shift();
            for (let neighbor of adjacencyList[node]) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) q.push(neighbor);
            }
            order.push(node);
        }
        return order.length === numCourses ? order : []
    }
}

File: Data-Structure/Alg-Study-Plan/课程表/00210. Course Schedule II/210. Course Schedule II.java
/*
基本思路
直观地说，拓扑排序就是让你把一幅无环图「拉平」，而且这个「拉平」的图里面，所有箭头方向都是一致的：

在进行拓扑排序之前，首先要确保图中无环，这就依赖 207. 课程表 中讲的环检测算法。

拓扑排序可以使用 DFS 算法，图的后序遍历结果进行反转就是拓扑排序结果。

另外，也可以用 BFS 算法借助每个节点的入度进行拓扑排序，这里就用 BFS 算法来解决。

DFS 解法和算法执行过程详解请看详细题解。

详细题解：拓扑排序详解及应用

标签：数据结构，图论算法，拓扑排序

解法代码
*/
class Solution {
    // 主函数
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        // 建图，和环检测算法相同
        List<Integer>[] graph = buildGraph(numCourses, prerequisites);
        // 计算入度，和环检测算法相同
        int[] indgree = new int[numCourses];
        for (int[] edge : prerequisites) {
            int from = edge[1], to = edge[0];
            indgree[to]++;
        }

        // 根据入度初始化队列中的节点，和环检测算法相同
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indgree[i] == 0) {
                q.offer(i);

            }
        }

        // 记录拓扑排序结果
        int[] res = new int[numCourses];
        // 记录遍历节点的顺序（索引）
        int count = 0;
        // 开始执行 BFS 算法
        while (!q.isEmpty()) {
            int cur = q.poll();
            // 弹出节点的顺序即为拓扑排序结果
            res[count] = cur;
            count++;
            for (int next : graph[cur]) {

                indgree[next]--;
                if (indgree[next] == 0) {
                    q.offer(next);
                }
            }
        }

        if (count != numCourses) {
            // 存在环，拓扑排序不存在
            return new int[] {};
        }

        return res;
    }

    // 建图函数
    List<Integer>[] buildGraph(int numCourses, int[][] prerequisites) {
        // 图中共有 numCourses 个节点
        List<Integer>[] graph = new LinkedList[numCourses];
        for (int i = 0; i < numCourses; i++) {
            graph[i] = new LinkedList<>();
        }
        for (int[] edge : prerequisites) {
            int from = edge[1], to = edge[0];
            // 修完课程 from 才能修课程 to
            // 在图中添加一条从 from 指向 to 的有向边
            graph[from].add(to);
        }
        return graph;
    }
}
/*
类似题目：

207. 课程表
*/